File: reality/seed/present/voices/llm/ssrf.js
// TreeOS Seed . AGPL-3.0 . https://treeos.ai . Tabor Holly
//
// ssrf.js — network safety for LLM connection base URLs.
//
// Every base URL operators (or extensions) hand to addLlmConnection
// passes through here. The point is keeping a misconfigured or
// hostile URL from reaching internal services (cloud metadata, local
// admin panels, the place's own loopback). We block:
//
//   - localhost / 0.0.0.0 / ::1 by hostname
//   - this reality's own URL (so an LLM connection can't ping our
//     own admin surface)
//   - cloud-provider metadata endpoints (169.254.169.254 et al.)
//   - private IP ranges (RFC 1918, link-local, CGNAT, unique-local,
//     IPv6 link-local, ULA)
//   - URLs with embedded credentials, non-http(s) protocols
//
// Opt-in path: `allowedLlmDomains` in reality config. Hosts on that
// list bypass the SSRF gate — that's how an operator stands up
// Ollama on the LAN or a self-hosted LLM gateway. Without that
// opt-in, the gate stays closed.
//
// DNS is resolved at validation time, not at call time, so a host
// that resolves to a private IP today can't be approved with a
// promise to "use only public addresses later." If the host's
// resolution changes after the connection is created, the moment-
// time call still hits the resolved IP and any subsequent re-
// validate would catch a regression.

import dns from "dns/promises";
import { getInternalConfigValue } from "../../../internalConfig.js";
import { getRealityConfigValue } from "../../../realityConfig.js";

const BLOCKED_HOSTS = new Set([
  "localhost",
  "0.0.0.0",
  "[::1]",
  "metadata.google.internal",
  "169.254.169.254",
  "metadata.internal",
]);

// Add this reality's own hostname to the block list at module load
// so a misconfigured LLM connection can't ping our admin surface.
try {
  const realityUrl = getRealityConfigValue("realityUrl");
  if (realityUrl) BLOCKED_HOSTS.add(new URL(realityUrl).hostname);
} catch {}

const BLOCKED_IP_PATTERNS = [
  /^127\./,
  /^10\./,
  /^192\.168\./,
  /^172\.(1[6-9]|2[0-9]|3[01])\./,
  /^169\.254\./,
  /^0\./,
  /^100\.(6[4-9]|[7-9]\d|1[0-2]\d)\./,
  /^198\.18\./,
  /^198\.19\./,
  /^fc/i,
  /^fe80/i,
  /^::1$/,
  /^::$/,
];

function isBlockedIp(ip) {
  return BLOCKED_IP_PATTERNS.some((p) => p.test(ip));
}

/**
 * Resolve a hostname and refuse if any A/AAAA record points at a
 * private/internal IP. Plain-IP hostnames check the IP directly.
 * DNS has a hard 5-second deadline; a hang here would block
 * connection setup indefinitely.
 */
export async function resolveAndValidateHost(hostname) {
  if (/^\d+\.\d+\.\d+\.\d+$/.test(hostname)) {
    if (isBlockedIp(hostname)) {
      throw new Error("URL points to a private/internal IP");
    }
    return;
  }

  try {
    const result = await Promise.race([
      dns.lookup(hostname, { all: true }),
      new Promise((_, reject) =>
        setTimeout(
          () => reject(new Error("DNS lookup timed out")),
          Number(getInternalConfigValue("dnsLookupTimeout")) || 5000,
        ),
      ),
    ]);
    for (const entry of result) {
      if (isBlockedIp(entry.address)) {
        throw new Error("URL resolves to a private/internal IP");
      }
    }
  } catch (err) {
    if (err.message.includes("private") || err.message.includes("internal")) {
      throw err;
    }
    throw new Error(
      "Could not resolve hostname: " +
        hostname +
        (err.message.includes("timed out") ? " (DNS timeout)" : ""),
    );
  }
}

/**
 * Is this hostname (or any superdomain of it) on the place's
 * `allowedLlmDomains` config list? Returns false when the list is
 * absent or empty — without an explicit opt-in, no host gets the
 * SSRF bypass.
 */
export function hostInAllowedLlmDomains(hostname) {
  const allowed = getRealityConfigValue("allowedLlmDomains");
  if (!Array.isArray(allowed) || allowed.length === 0) return false;
  return allowed.some((d) => {
    const low = d.toLowerCase();
    return hostname === low || hostname.endsWith("." + low);
  });
}

/**
 * Validate an LLM connection's base URL. Returns the canonicalized
 * URL (no trailing slash) on success; throws otherwise.
 *
 *   - protocol must be http(s)
 *   - no embedded credentials
 *   - host on `allowedLlmDomains` → bypass everything else
 *   - blocked hostnames refuse
 *   - private/internal IPs refuse (without allowedLlmDomains opt-in)
 *   - when allowedLlmDomains is set, any host NOT on it refuses
 *
 * Note: this is the SYNCHRONOUS validation. The async DNS resolution
 * check (`resolveAndValidateHost`) runs separately in the CRUD path
 * so the connection setup can refuse late-bound private IPs.
 */
export function validateBaseUrl(baseUrl) {
  let parsed;
  try {
    parsed = new URL(baseUrl);
  } catch {
    throw new Error("Invalid base URL");
  }
  if (!["http:", "https:"].includes(parsed.protocol)) {
    throw new Error("Only http and https URLs are allowed");
  }
  if (parsed.username || parsed.password) {
    throw new Error("URLs with credentials are not allowed");
  }
  const hostname = parsed.hostname.toLowerCase();
  if (hostInAllowedLlmDomains(hostname)) {
    return parsed.href.replace(/\/+$/, "");
  }
  if (BLOCKED_HOSTS.has(hostname)) {
    throw new Error("This base URL is not allowed");
  }
  if (isBlockedIp(hostname)) {
    throw new Error(
      "Local/private network URLs are not allowed. Add the host to " +
        "`allowedLlmDomains` in place config to opt in (e.g. for Ollama " +
        "or a LAN-hosted LLM).",
    );
  }
  const allowed = getRealityConfigValue("allowedLlmDomains");
  if (Array.isArray(allowed) && allowed.length > 0) {
    throw new Error(
      `LLM domain "${hostname}" is not in this reality's allowed list.`,
    );
  }
  return parsed.href.replace(/\/+$/, "");
}

File: rust/treecognition/src/ssrf.rs
// treecognition::ssrf — network safety for an LLM connection's base URL (port of ssrf.js). Keeps a
// misconfigured/hostile URL from reaching internal services (cloud metadata, local admin, loopback,
// the LAN). Blocks: localhost / 0.0.0.0 / ::1 by host; cloud metadata endpoints; private IP ranges
// (RFC1918, link-local, CGNAT, ULA, IPv6 link-local); credentials in the URL; non-http(s) schemes.
//
// Opt-in: a host (or any superdomain) on the story's `allowedLlmDomains` bypasses the gate — that's how
// an operator stands up Ollama on the LAN. Without it, the gate stays closed.
//
// `validate_base_url` is the SYNCHRONOUS check (pure). The DNS resolution check is I/O, so the lookup
// itself happens at the edge; `check_resolved_ips` is the pure rule it applies to each resolved A/AAAA.
// (Faithful to the JS, with one tightening: the IPv6 private-range checks only fire on colon-bearing
// hosts, so a domain like "fcserver.com" isn't mis-flagged the way the JS `/^fc/` regex would.)

/// Hosts refused outright. The story's own host is appended at the call site (`own_host`).
pub const BLOCKED_HOSTS: &[&str] = &["localhost", "0.0.0.0", "[::1]", "metadata.google.internal", "169.254.169.254", "metadata.internal"];

/// Is this address in a private/internal range? IPv4 prefixes (RFC1918 / link-local / CGNAT / TEST-NET)
/// plus IPv6 ULA / link-local / loopback / unspecified.
pub fn is_blocked_ip(host: &str) -> bool {
    let low = host.to_lowercase();
    // IPv4 literal prefixes
    if low.starts_with("127.") || low.starts_with("10.") || low.starts_with("192.168.") || low.starts_with("169.254.") || low.starts_with("0.") || low.starts_with("198.18.") || low.starts_with("198.19.") {
        return true;
    }
    // 172.16.0.0 – 172.31.255.255
    if let Some(rest) = low.strip_prefix("172.") {
        if let Ok(o) = rest.split('.').next().unwrap_or("").parse::<u32>() {
            if (16..=31).contains(&o) {
                return true;
            }
        }
    }
    // 100.64.0.0/10 (CGNAT)
    if let Some(rest) = low.strip_prefix("100.") {
        if let Ok(o) = rest.split('.').next().unwrap_or("").parse::<u32>() {
            if (64..=127).contains(&o) {
                return true;
            }
        }
    }
    // IPv6 (strip [] brackets; only colon-bearing hosts, so a "fc…" DOMAIN isn't mis-flagged)
    let v6 = low.trim_start_matches('[').trim_end_matches(']');
    if v6.contains(':') && (v6.starts_with("fc") || v6.starts_with("fd") || v6.starts_with("fe80") || v6 == "::1" || v6 == "::") {
        return true;
    }
    false
}

/// Is `hostname` (or a superdomain) on the opt-in `allowed` list? Empty list = no bypass (gate closed).
pub fn host_in_allowed_llm_domains(hostname: &str, allowed: &[String]) -> bool {
    allowed.iter().any(|d| {
        let low = d.to_lowercase();
        hostname == low || hostname.ends_with(&format!(".{low}"))
    })
}

/// Apply the private-IP rule to each DNS-resolved address (the edge resolves; this is the pure gate).
/// Err on the first private/internal hit, so a public hostname that resolves inward is refused.
pub fn check_resolved_ips(ips: &[String]) -> Result<(), String> {
    for ip in ips {
        if is_blocked_ip(ip) {
            return Err("URL resolves to a private/internal IP".to_string());
        }
    }
    Ok(())
}

/// Validate an LLM base URL synchronously -> the canonicalized URL (no trailing slash), or an error.
/// `own_host` is the story's own hostname (also refused). `allowed` is the opt-in domain list.
pub fn validate_base_url(base_url: &str, allowed: &[String], own_host: Option<&str>) -> Result<String, String> {
    let (scheme, rest) = base_url.split_once("://").ok_or("Invalid base URL")?;
    match scheme.to_lowercase().as_str() {
        "http" | "https" => {}
        _ => return Err("Only http and https URLs are allowed".to_string()),
    }
    // authority = up to the first path/query/fragment delimiter
    let authority = &rest[..rest.find(['/', '?', '#']).unwrap_or(rest.len())];
    if authority.contains('@') {
        return Err("URLs with credentials are not allowed".to_string());
    }
    // hostname = authority minus :port (keep [..] for IPv6)
    let hostname = if let Some(stripped) = authority.strip_prefix('[') {
        // [ipv6](:port)? -> keep the bracketed form
        let end = stripped.find(']').map(|i| i + 1).unwrap_or(stripped.len());
        format!("[{}", &stripped[..end])
    } else {
        authority.split(':').next().unwrap_or(authority).to_string()
    }
    .to_lowercase();
    if hostname.is_empty() {
        return Err("Invalid base URL".to_string());
    }

    if host_in_allowed_llm_domains(&hostname, allowed) {
        return Ok(base_url.trim_end_matches('/').to_string());
    }
    if BLOCKED_HOSTS.contains(&hostname.as_str()) || own_host.is_some_and(|h| h.eq_ignore_ascii_case(&hostname)) {
        return Err("This base URL is not allowed".to_string());
    }
    if is_blocked_ip(&hostname) {
        return Err("Local/private network URLs are not allowed. Add the host to `allowedLlmDomains` in story config to opt in (e.g. for Ollama or a LAN-hosted LLM).".to_string());
    }
    if !allowed.is_empty() {
        return Err(format!("LLM domain \"{hostname}\" is not in this story's allowed list."));
    }
    Ok(base_url.trim_end_matches('/').to_string())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn no_allow() -> Vec<String> {
        vec![]
    }

    #[test]
    fn blocks_private_and_metadata_and_loopback() {
        assert!(is_blocked_ip("127.0.0.1"));
        assert!(is_blocked_ip("10.1.2.3"));
        assert!(is_blocked_ip("192.168.0.1"));
        assert!(is_blocked_ip("172.16.5.5") && is_blocked_ip("172.31.0.1"));
        assert!(!is_blocked_ip("172.32.0.1")); // just outside the RFC1918 /12
        assert!(is_blocked_ip("169.254.1.1")); // link-local / metadata
        assert!(is_blocked_ip("100.64.0.1") && !is_blocked_ip("100.63.0.1")); // CGNAT edge
        assert!(is_blocked_ip("[fc00::1]") && is_blocked_ip("fe80::1") && is_blocked_ip("::1"));
        assert!(!is_blocked_ip("fcserver.com")); // a domain, not an IPv6 ULA
        assert!(!is_blocked_ip("8.8.8.8"));
    }

    #[test]
    fn validate_refuses_the_dangerous_and_passes_public() {
        assert!(validate_base_url("ftp://x.com", &no_allow(), None).is_err()); // scheme
        assert!(validate_base_url("http://user:pw@api.openai.com", &no_allow(), None).is_err()); // creds
        assert!(validate_base_url("http://localhost:11434", &no_allow(), None).is_err()); // blocked host
        assert!(validate_base_url("http://169.254.169.254/latest", &no_allow(), None).is_err()); // metadata IP
        assert!(validate_base_url("http://10.0.0.5:8080", &no_allow(), None).is_err()); // private IP
        // a public endpoint passes and is canonicalized (no trailing slash)
        assert_eq!(validate_base_url("https://api.openai.com/v1/", &no_allow(), None).unwrap(), "https://api.openai.com/v1");
    }

    #[test]
    fn allowed_domains_opt_in_bypasses_and_closes_others() {
        let allow = vec!["lan.example".to_string()];
        // a LAN host opted in passes despite being private
        assert!(validate_base_url("http://ollama.lan.example:11434", &allow, None).is_ok());
        // with a non-empty allow list, anything NOT on it is refused (even public)
        assert!(validate_base_url("https://api.openai.com", &allow, None).is_err());
    }

    #[test]
    fn own_story_host_is_refused() {
        assert!(validate_base_url("https://my.story", &no_allow(), Some("my.story")).is_err());
    }

    #[test]
    fn resolved_private_ip_is_caught() {
        assert!(check_resolved_ips(&["8.8.8.8".into(), "10.0.0.1".into()]).is_err());
        assert!(check_resolved_ips(&["8.8.8.8".into(), "1.1.1.1".into()]).is_ok());
    }
}
